traits/trait_set_object.py

```diff
@@ -51,213 +50,518 @@ def __repr__(self):
+
+        This simply calls all notifiers provided by the class, if any.
+        The notifiers are expected to have the signature::
+
+            notifier(removed, added)
+
+        Any return values are ignored. Any exceptions raised are not
+        handled. Notifiers are therefore expected not to raise any
+        exceptions under normal use.
+
+        Parameters
+        ----------
+        removed : set
+            The items that have been removed.
+        added : set
+            The new items that have been added to the set.
+        """
+        for notifier in self.notifiers:
+            notifier(removed, added)
+
+    # -- set interface -------------------------------------------------------
+
+    def __iand__(self, value):
+        """  Return self &= value.
+
+        Parameters
+        ----------
+        value : set or frozenset
+            A value.
+
+        Returns
+        -------
+        self : TraitSet
+            The updated set.
+        """
+
+        old_set = self.copy()
+        retval = super().__iand__(value)
+        removed = old_set.difference(self)
+
+        if len(removed) > 0:
+            self.notify(removed, set())
+
+        return retval
+
+    def __ior__(self, value):
+        """ Return self |= value.
+
+        Parameters
+        ----------
+        value : set or frozenset
+            A value.
+
+        Returns
+        -------
+        self : TraitSet
+            The updated set.
+        """
+        old_set = self.copy()
+
+        if isinstance(value, (set, frozenset)):
+            value = {self.item_validator(item)
+                     for item in value}
+
+        retval = super().__ior__(value)
+
+        added = self.difference(old_set)
+
+        if len(added) > 0:
+            self.notify(set(), added)
+
+        return retval
+
+    def __isub__(self, value):
+        """ Return self-=value.
+
+        Parameters
+        ----------
+        value : set or frozenset
+            A value.
+
+        Returns
+        -------
+        self : TraitSet
+            The updated set.
+        """
+
+        old_set = self.copy()
+        retval = super().__isub__(value)
+        removed = old_set.difference(self)
+
+        if len(removed) > 0:
+            self.notify(removed, set())
+
+        return retval
+
+    def __ixor__(self, value):
+        """ Return self ^= value.
+
+        Parameters
+        ----------
+        value : set or frozenset
+            A value.
+
+        Returns
+        -------
+        self : TraitSet
+            The updated set.
+        """
+
+        removed = set()
+        added = set()
+        if isinstance(value, (set, frozenset)):
+            values = set(value)
+            removed = self.intersection(values)
+            raw_result = values.difference(removed)
+            validated_result = {self.item_validator(item) for item in
+                                raw_result}
+            added = validated_result.difference(self)
```

Comment:
I'm not sure I'm following the logic here. Why is the `difference` call needed? Shouldn't `validated_result` already be disjoint from `self`?

Reply:
Okay, I see; this matters when the validator is transforming values, and the transformed values aren't equal to the originals.

Off-topic for this PR, but I think we should probably document somewhere that we simply don't support this case, and any behaviour should be treated as accidental. It's almost impossible to come up with reasonable and consistent semantics for the various operations.

(In contrast, the case where the item validator may transform the value, but the transformed value compares _equal_ to the original is reasonable: for example, the way that the `Int` trait converts NumPy ints to regular Python ints.)

And now I'm worrying about the case of `Float` traits transforming `Int`s, which _is_ a potential value-converting transformation.

One possible solution is to require/document that `item_validators` should be idempotent. We may then be able to get away with always validating inputs.

This may need more thought, but that shouldn't block this PR.

Reply:
Opened #1020 for discussion.